pkg/storegateway/indexheader/threadpool.go

```diff
@@ -0,0 +1,112 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/cpg1111/threadpool-go/blob/master/threadpool.go
+// Provenance-includes-license: MIT
+// Provenance-includes-copyright: Christian Grabowski
+
+package indexheader
+
+import (
+	"errors"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	LabelWaiting  = "waiting"
+	LabelComplete = "complete"
+)
+
+var ErrPoolStopped = errors.New("thread pool has been stopped")
+
+type Threadpool struct {
+	pool       chan *OSThread
+	stopping   chan struct{}
+	stopped    chan struct{}
+	numThreads int
+
+	timing *prometheus.HistogramVec
+	tasks  prometheus.Gauge
+}
+
+func NewThreadPool(num int, reg prometheus.Registerer) (*Threadpool, error) {
+	if num <= 0 {
+		return nil, nil
+	}
+
+	tp := &Threadpool{
+		pool:       make(chan *OSThread, num),
+		stopping:   make(chan struct{}),
+		stopped:    make(chan struct{}),
+		numThreads: num,
+
+		timing: promauto.With(reg).NewHistogramVec(prometheus.HistogramOpts{
```

Comment:
I miss the buckets here, and I think that no buckets will work well here: "waiting stage" will be something moving in milliseconds, while "complete" is something that depends a lot on a task and can be even seconds. So this will end up with a huge cardinality on the buckets.

I'd consider using a summary for the waiting time (exporting the p99, then we can graph the max of p99), and maybe (not sure about this) even dropping the "complete" stage at all since it's something out of this pool's control